src/dataframe/dataframe.jl

```diff
@@ -394,65 +393,78 @@ end
 """
     setindex!(df::DataFrame, v, col_ind::Union{Integer, Symbol})
 
-If `v` is not an `AbstractVector` then column `col_ind` is filled with `v`.
-Otherwise `v` is used as a column vector (no data copying is performed).
-`col_ind` can be either a `Symbol` or `Integer` (except `Bool`).
-If `col_ind` is present in a `DataFrame` then:
-* if `v` is an abstract vector then it replaces `df[col_ind]`
-* if `v` i not an abstract vector then existing `df[col_ind]` is filled with `v`
+Assign `v` to column `col_ind` of data frame `df`.
+
+If column `col_ind` is already present in `df` then:
+* if `v` is an `AbstractVector` then it replaces `df[col_ind]` (no copy is made)
+* otherwise, the existing `df[col_ind]` vector is filled with `v`
   (which will fail if `v` cannot be converted to `eltype(df[col_ind])`)
-If `col_ind` is not present in a `DataFrame` then an appropriate column is added to it.
-In particular you can add a column number `ncol(df)+1` to a `DataFrame` this way.
 
+If column `col_ind` does not exist in `df` then a new column is added at the end of `df` and:
+* if `v` is an `AbstractVector`, it is used as a column (no copy is made).
+* otherwise, a `Vector` is created by repeating `v` as many times as there are rows in `df`.
+
+As a special case, `df[ncol(df)+1] = v` is allowed and creates a new column at the end
+of `df`.
+"""
+Base.setindex!(df::DataFrame, v, col_ind)
+
+"""
     setindex!(df::DataFrame, v, col_inds::AbstractVector})
 
-`col_inds` can be either vector of `Symbol`, `Integer`, `Union{Bool, Missing}` or `:`.
+Assign `v` to columns `col_inds` of data frame `df`.
 
-If `v` is not a `DataFrame` then assignment rules follow single column `setindex!`.
+`col_inds` can be either `:`, a vector of `Symbol` or `Integer`, or a vector of `Bool`
+of the same length as the number of columns in `df` (logical indexing).
 
-If `v` is a `DataFrame` then assignment to `col_inds` take columns from `v` sequentially
-disregarding their names.  Number of rows of `df` and `v` must be identical in this case.
-Additionally number of columns in `v` must be equal to length of `col_inds`.
+If `v` is not a data frame then assignment rules follow single-column `setindex!` .
 
-You can create new columns in `df` this way and the rules are the same as for `setindex!`
-on a single column.
+If `v` is a data frame then each column in `v` is assigned sequentially to the column
+at the corresponding position in `col_inds`, disregarding column names.
+`df` and `v` must have the same number of rows, and the number of columns in `v`
+must be equal to the length of `col_inds`.
 
-If `col_inds<:AbstractVector{Union{Bool, Missing}}` then it must only contain `Bool` values
```

Comment:
I've removed this because it would be long to explain that we don't care about the actual type of vectors, only about their contents, except for logical indexing, and because mentioning `Missing` could be misinterpreted as saying that missing values are accepted. Probably better simply mention the type of the actual elements.